### DISCUSSION TITLE
Can't create alpha prerelease with new major version from another prerelease branch which is beta

### DISCUSSION THREAD

**Discussion by nmihalyov** [semantic-release/semantic-release#2386] [Release workflow]
## Current behavior

I have version prerelease branch with version 8.0.0-beta.x and trying to create new release with version 9.0.0-alpha.x by creating new branch *version/9* from *version/9*, but semantic-release creates version 8.0.0-alpha.x

## Expected behavior

Semantic-release creates 9.0.0-alpha.x version on new branch that is created from other prerelease branch.

## Environment

- **semantic-release** version: 18.0.0
- CI environment: GitLab CI
- Plugins used: @semantic-release/commit-analyzer, @semantic-release/release-notes-generator, @semantic-release/changelog, @semantic-release/npm, @semantic-release/git
- **semantic-release** configuration:
```
const NOTE_KEYWORDS = ["BREAKING CHANGE", "BREAKING CHANGES"];
const REVERT_PATTERN = /^(?:Revert|revert:)\s"?([\s\S]+?)"?\s*this reverts commit (\w*)/i;

module.exports = {
  "branches": [
    {
      "name": "master"
    },
    {
      "name": "version/8",
      "prerelease": "beta"
    },
    {
      "name": "version/9",
      "prerelease": "alpha"
    }
  ],
  "plugins": [
    ["@semantic-release/commit-analyzer", {
      "preset": "angular",
      "parserOpts": {
        "noteKeywords": NOTE_KEYWORDS,
        "revertPattern": REVERT_PATTERN,
      }
    }],
    ["@semantic-release/release-notes-generator", {
      "preset": "angular",
      "parserOpts": {
        "noteKeywords": NOTE_KEYWORDS,
        "revertPattern": REVERT_PATTERN,
      }
    }],
    [
      "@semantic-release/changelog",
      {
        "changelogFile": "docs/CHANGELOG.md"
      }
    ],
    "@semantic-release/npm",
    [
      "@semantic-release/git",
      {
        "assets": [
          "docs/CHANGELOG.md",
          "package.json",
          "package-lock.json"
        ]
      }
    ]
  ]
}
```
- CI logs:
```
[8:01:23 AM] [semantic-release] › ℹ  Found git tag v7.348.0 associated with version 7.348.0 on branch version/9
[8:01:23 AM] [semantic-release] › ℹ  Found [71](url) commits since last release
[8:01:23 AM] [semantic-release] › ℹ  Start step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[8:01:23 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: feat(ci): set up environment for 9th version
BREAKING CHANGE: new release
[8:01:23 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The release type for the commit is major
[8:01:23 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analysis of 71 commits complete: major release
[8:01:23 AM] [semantic-release] › ✔  Completed step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[8:01:23 AM] [semantic-release] › ℹ  The next release version is 8.0.0-alpha.1
```

- Git history:
<img width="1064" alt="image" src="https://user-images.githubusercontent.com/15937747/156724260-01e86a3c-6982-4aee-8e0a-28a8252b033f.png">

**Comment by travi** [comment]
this is working as intended. semantic-release works by determining the significance of changes since the previous release. for pre-releases, that process goes further by comparing to the previous stable release. 

when working from a pre-release branch that is created from another pre-release branch, the previous stable release that exists in the new branch is the same as the one from the previous pre-release branch, so the new version is determined based on the changes since that existing stable release. Even with multiple breaking changes made as pre-releases, per semver, the major version should only be incremented once. 

If you want it to be increased multiple times, you need to make a stable release and ensure that that release exists in the history of the pre-release branch that semantic-release is operating upon.

**Reply by nmihalyov** [reply]
Thanks for the answer! But could you please correct me where I understood it wrong 'cause by [this guide](https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/release-workflow/pre-releases.md#working-on-another-future-release) I thought we could just create 3.0.0-alpa from 2.0.0-beta without releasing it as a stable version

**Reply by travi** [reply]
sorry for the delay. yes, that example is incorrect, as recently discovered in another discussion: https://github.com/semantic-release/semantic-release/discussions/2358. you can find more details about the reasons in that discussion.

if you would be open to helping us update that example, or decide if it should simply be removed, that could be helpful.

**Comment by yauheni-zmitrovich-w** [comment]
Hi @travi.

If the behavior is expected could you please confirm that the [Pre-releases Recipe](https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/release-workflow/pre-releases.md) is incorrect?

Graph from the documentation
```
1  * feat: initial commit # => v1.0.0 on @latest
2  | \
3  |  * feat: first feature \n\n BREAKING CHANGE: it breaks something # => v2.0.0-beta.1 on @beta
4  |  * feat: second feature # => v2.0.0-beta.2 on @beta
5  *  | fix: a fix # => v1.0.1 on @latest
6  |  | \
7  |  |  * feat: first feature of other release \n\n BREAKING CHANGE: it breaks something # => v3.0.0-alpha.1 on @alpha
8  |  |  * feat: second feature of other release # => v3.0.0-alpha.2 on @alpha
9  | /|  |
10 *  |  | Merge branch beta into master # => v2.0.0 on @latest
11 |  | /|
12 |  *  | Merge branch alpha into beta # => v3.0.0-beta.1 on @beta
13 | /|  |
14 *  |  | Merge branch beta into master # => v3.0.0 on @latest
```

Should the graph be fixed as described below if the behavior is expected?
```
1  * feat: initial commit # => v1.0.0 on @latest
2  | \
3  |  * feat: first feature \n\n BREAKING CHANGE: it breaks something # => v2.0.0-beta.1 on @beta
4  |  * feat: second feature # => v2.0.0-beta.2 on @beta
5  *  | fix: a fix # => v1.0.1 on @latest
6  |  | \
7  |  |  * feat: first feature of other release \n\n BREAKING CHANGE: it breaks something # => v2.0.0-alpha.1 on @alpha
8  |  |  * feat: second feature of other release # => v2.0.0-alpha.2 on @alpha
9  | /|  |
10 *  |  | Merge branch beta into master # => v2.0.0 on @latest
11 | \|  | 
12 |  *  | Merge branch master into beta
13 |  | /|
14 |  *  | Merge branch alpha into beta # => v3.0.0-beta.1 on @beta
15 | /|  |
16 *  |  | Merge branch beta into master # => v3.0.0 on @latest
```
Changed v3 to v2 on lines 7,8
Added step `Merge branch master into beta` on lines 11-13

What are the correct graph and expected behavior?

Thanks in advance!

**Reply by travi** [reply]
yes, that example is incorrect, as recently discovered in another discussion: https://github.com/semantic-release/semantic-release/discussions/2358. you can find more details about the reasons in that discussion.

in your suggested update, i would suggest that once the beta branch has been merged into the mainline, that pre-release should be considered complete and the beta branch should be deleted. in order for the alpha branch to be aware of the previous beta being promoted to stable, the mainline should be merged into the alpha branch. at that point, it could be promoted to beta by creating beta branch from the alpha branch, or left as alpha if desired. 

to be honest, having multiple pre-releases that depend upon each other is not a scenario that should be useful very often. instead, it is usually better to keep them independent and create each from the mainline. do you have a concrete scenario where this workflow was needed?

if you would be open to helping us update that example, or decide if it should simply be removed, that could be helpful.

**Comment by taco-chen** [comment]
Thank you to @travi for the detailed explanation.

I have encountered the same issue as the original post. In our scenario, we typically release a prerelease beta version for internal testing, which includes some features and bug fixes. Concurrently, our team advances to the alpha version for our next release. Our situation precisely matches the example in the [documentation](https://semantic-release.gitbook.io/semantic-release/recipes/release-workflow/pre-releases#publishing-the-2.0.0-beta-release-to-the-default-distribution-channel).

```
* feat: initial commit # => v1.0.0 on @latest
| \
|  * feat: first feature \n\n BREAKING CHANGE: it breaks something # => v2.0.0-beta.1 on @beta
|  * feat: second feature # => v2.0.0-beta.2 on @beta
*  | fix: a fix # => v1.0.1 on @latest
|  | \
|  |  * feat: first feature of other release \n\n BREAKING CHANGE: it breaks something # => v3.0.0-alpha.1 on @alpha
|  |  * feat: second feature of other release # => v3.0.0-alpha.2 on @alpha
| /|  |
*  |  | Merge branch beta into master/main # => v2.0.0 on @latest
```

In our case, incrementing the alpha version number based on the beta is crucial for maintaining multiple releases simultaneously. For example, we might have:

1. 2.0.0-beta.1 for feature A, which will be released soon.
2. 3.0.0-alpha.1 for feature B, which will be released later.

The increased alpha version number is very useful for our developers and testers to distinguish that 2.0.0-beta.1 and 3.0.0-alpha.1 are entirely different versions (although 2.0.0-beta.1 should be an ancestor of 3.0.0-alpha.1).

Based on the discussion in https://github.com/semantic-release/semantic-release/discussions/2358, I understand that independence between different prerelease versions is important in some cases. However, in other cases like ours, it would be greatly appreciated if we could have a feature similar to what is depicted in the [documentation](https://semantic-release.gitbook.io/semantic-release/recipes/release-workflow/pre-releases#publishing-the-2.0.0-beta-release-to-the-default-distribution-channel).

## My current workaround

To maintain both the workflow described in the [documentation](https://semantic-release.gitbook.io/semantic-release/recipes/release-workflow/pre-releases#publishing-the-2.0.0-beta-release-to-the-default-distribution-channel) and the independence discussed in https://github.com/semantic-release/semantic-release/discussions/2358, my current workaround is to extend the `.releaserc.json` file as shown below:

```.releaserc.json
  "branches": [
    "main",
    {
      "name": "beta",
      "prerelease": true
    },
    {
      "name": "alpha",
      "prerelease": true,
      "baseBranch": "beta"
    }
  ],
```

If a branch has its configuration with baseBranches, the version number will consider the Base Branch as a normal release branch:

```
* feat: initial commit # => v1.0.0 on @latest
| \
|  * feat: first feature \n\n BREAKING CHANGE: it breaks something # => v2.0.0-beta.1 on @beta
|  * feat: second feature # => v2.0.0-beta.2 on @beta
*  | fix: a fix # => v1.0.1 on @latest
|  | \
|  |  * feat: first feature of other release \n\n BREAKING CHANGE: it breaks something # => v3.0.0-alpha.1 on @alpha
```

However, since I am unable to find a suitable entry point in the steps listed in [@semantic-release/exec](https://github.com/semantic-release/exec), I use an additional shell script to achieve this. My workaround involves:

1. Checking if we are on a prerelease branch with `baseBranch` configuration.
2. Temporarily promoting the base branch's prerelease version number.
3. Running `$ npx semantic-release.`
4. Removing all promoted tags.

Example script output:
```
You are on the 'alpha' branch.
Base hash: 52e0bbd found for version 1.0.0.
Promoting preprelease version with 'beta' tags.
  - Tag 2.0.0-beta.1 has been promoted to 2.0.0 on 751c7af.
...
[1:35:17 pm] [semantic-release] › ℹ  The next release version is 3.0.0-alpha.1
...
Cleaning up promoted develop tags.
  - Deleted tag '2.0.0' (was 751c7af)
```

## Feature Suggestion

I am wondering if we could have this feature officially supported. My proposal using `baseBranch` may not be perfect, but it would be incredibly helpful for teams like ours to have this simulation feature. I am more than happy to contribute in any way I can.

Example `.prerelease.json` configuration:

```.prerelease.json
  "branches": [
    "main",
    {
      "name": "beta",
      "prerelease": true
    },
    {
      "name": "alpha",
      "prerelease": true,
      "baseBranch": "beta"
    }
  ],
```

Thank you ❤️ !
